pkg/runtime/framework/plugins/coscheduling/coscheduling.go

```diff
@@ -121,41 +117,23 @@ func (c *CoScheduling) Build(ctx context.Context, _ client.Object, info *runtime
 			totalResources[resName] = current
 		}
 	}
-	newPG := &schedulerpluginsv1alpha1.PodGroup{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: schedulerpluginsv1alpha1.SchemeGroupVersion.String(),
-			Kind:       constants.PodGroupKind,
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      trainJob.Name,
-			Namespace: trainJob.Namespace,
-		},
-		Spec: schedulerpluginsv1alpha1.PodGroupSpec{
-			ScheduleTimeoutSeconds: info.RuntimePolicy.PodGroupPolicy.Coscheduling.ScheduleTimeoutSeconds,
-			MinMember:              totalMembers,
-			MinResources:           totalResources,
-		},
-	}
-	if err := ctrlutil.SetControllerReference(trainJob, newPG, c.scheme); err != nil {
-		return nil, err
-	}
-	oldPG := &schedulerpluginsv1alpha1.PodGroup{}
-	if err := c.client.Get(ctx, client.ObjectKeyFromObject(newPG), oldPG); err != nil {
-		if !apierrors.IsNotFound(err) {
-			return nil, err
-		}
-		oldPG = nil
-	}
-	if needsCreateOrUpdate(oldPG, newPG, ptr.Deref(trainJob.Spec.Suspend, false)) {
-		return []client.Object{newPG}, nil
-	}
-	return nil, nil
-}
 
-func needsCreateOrUpdate(old, new *schedulerpluginsv1alpha1.PodGroup, trainJobIsSuspended bool) bool {
-	return old == nil ||
-		trainJobIsSuspended &&
-			(!equality.Semantic.DeepEqual(old.Spec, new.Spec) || !maps.Equal(old.Labels, new.Labels) || !maps.Equal(old.Annotations, new.Annotations))
+	podGroup := schedulerpluginsv1alpha1ac.PodGroup(trainJob.Name, trainJob.Namespace)
+
+	podGroup.WithSpec(schedulerpluginsv1alpha1ac.PodGroupSpec().
+		WithMinMember(totalMembers).
+		WithMinResources(totalResources).
+		WithScheduleTimeoutSeconds(*info.RuntimePolicy.PodGroupPolicy.Coscheduling.ScheduleTimeoutSeconds))
+
+	podGroup.WithOwnerReferences(metav1ac.OwnerReference().
+		WithAPIVersion(trainer.GroupVersion.String()).
+		WithKind(trainer.TrainJobKind).
+		WithName(trainJob.Name).
+		WithUID(trainJob.UID).
+		WithController(true).
+		WithBlockOwnerDeletion(true))
+
+	return []any{podGroup}, nil
```

Comment:
Could you check if TrainJob is suspended before returning podGroup?
Because we do not want to fire reconcile trigger when trainjob is suspended.

Reply:
You're right, I've isolated the logic in the jobset plugin and that needs to be there as well.

Actually I remember I wanted to raise that point with you as it feels the wrong level to handle the case where the TrainJob is suspended. I'd intuitively move it into the TrainJob controller instead so it's global to all the components / plugins.

Reply:
>  I'd intuitively move it into the TrainJob controller instead so it's global to all the components / plugins.

It sounds reasonable. If we want to just check TrainJob, the checking should be implemented outside of pipeline.

Reply:
OK I realize this may be a bit more complex to move that part outside of the pipeline as the state of the JobSet for example needs to be checked to make sure its suspend field is synced with that of the TrainJob.
I'm going to keep the existing logic and think about it more.

Reply:
That makes sense. We can optimize it later.